compiler/src/dotty/tools/dotc/ast/Desugar.scala

```diff
@@ -699,7 +699,10 @@ object desugar {
           DefDef(methName, derivedTparams, (unapplyParam :: Nil) :: Nil, unapplyResTp, unapplyRHS)
             .withMods(synthetic)
         }
-        companionDefs(companionParent, applyMeths ::: unapplyMeth :: companionMembers)
+        val toStringMeth =
+          DefDef(nme.toString_, Nil, Nil, TypeTree(), Literal(Constant(className.toString))).withMods(Modifiers(Override | Synthetic))
```

Comment:
Scalac is a bit weird:
```scala
scala> case class Foo(x: Int); object Foo
class Foo
object Foo

scala> Foo
val res0: Foo.type = Foo$@fe34b86

scala> case class Foo(x: Int)
class Foo

scala> Foo
val res1: Foo.type = Foo
```
I'm not sure if it's worth replicating this exact behavior, but maybe worth checking the scalac logic which determines when toString is inserted.

Reply:
Seems like scalac only adds the toString in the situations where it also adds Function1 as a parent of the object (to avoid getting "<function1>" as the toString), it's a bit weird but doing the same would make it easier to do bytecode diff between scalac and dotty.

Reply:
There isn't a way to know if the module will extend `Function1` while desugaring. I can't find a place to do it while typing either. Any suggestions?

Reply:
Maybe check if the moduledef is compiler-generated (span.isZeroExtent) ?

Reply:
I could not find a tree here with a span that was different for explicit or non explicit modules.

Reply:
> There isn't a way to know if the module will extend Function1 while desugaring

Actually this is implemented in desugaring, see `companionParent`

Reply:
Unfortunately, both the get a `Function` in `companionParent`, the explicit declaration loses it somewhere later (not sure where).

Reply:
Any other idea?

Reply:
I think it's fine as is.